src/OpenTracing/Mock/MockSpan.cs

```diff
@@ -0,0 +1,306 @@
+        {
+            _mockTracer = tracer;
+            OperationName = operationName;
+            StartTimestamp = startTimestamp;
+
+            _tags = initialTags == null
+                ? new Dictionary<string, object>()
+                : new Dictionary<string, object>(initialTags);
+
+            _references = references == null
+                ? new List<Reference>()
+                : references.ToList();
+
+            var parentContext = FindPreferredParentRef(_references);
+
+            if (parentContext == null)
+            {
+                // we are a root span
+                _context = new MockSpanContext(NextId(), NextId(), new Dictionary<string, string>());
+                ParentId = 0;
+            }
+            else
+            {
+                // we are a child span
+                _context = new MockSpanContext(parentContext.TraceId, NextId(), MergeBaggages(_references));
+                ParentId = parentContext.SpanId;
+            }
+        }
+
+        public ISpan SetOperationName(string operationName)
+        {
+            CheckForFinished("Setting operationName [{0}] on already finished span", operationName);
+            OperationName = operationName;
+            return this;
+        }
+
+        public ISpan SetTag(string key, bool value)
+        {
+            return SetObjectTag(key, value);
+        }
+
+        public ISpan SetTag(string key, double value)
+        {
+            return SetObjectTag(key, value);
+        }
+
+        public ISpan SetTag(string key, int value)
+        {
+            return SetObjectTag(key, value);
+        }
+
+        public ISpan SetTag(string key, string value)
+        {
+            return SetObjectTag(key, value);
+        }
+
+        private ISpan SetObjectTag(string key, object value)
+        {
+            lock (_lock)
+            {
+                CheckForFinished("Setting tag [{0}:{1}] on already finished span", key, value);
+                _tags[key] = value;
+                return this;
+            }
+        }
+
+        public ISpan Log(IDictionary<string, object> fields)
+        {
+            return Log(DateTimeOffset.UtcNow, fields);
+        }
+
+        public ISpan Log(DateTimeOffset timestamp, IDictionary<string, object> fields)
+        {
+            lock (_lock)
+            {
+                CheckForFinished("Adding logs {0} at {1} to already finished span.", fields, timestamp);
+                _logEntries.Add(new LogEntry(timestamp, fields));
+                return this;
+            }
+        }
+
+        public ISpan Log(string @event)
+        {
+            return Log(DateTimeOffset.UtcNow, @event);
+        }
+
+        public ISpan Log(DateTimeOffset timestamp, string @event)
+        {
+            return Log(timestamp, new Dictionary<string, object> { { "event", @event } });
+        }
+
+        public ISpan SetBaggageItem(string key, string value)
+        {
+            lock (_lock)
```

Comment:
:( locks hurt

`System.Collections.Concurrent` avoid locks, and definitely minimize them (e.g. why is the string format inside the lock?)
This is fine if this is genuinely a MockTracer, only used for testing, but it almost seems like it's an InMemoryTracer. If we are not going to have a separate InMemoryTracer, then I'd suggest we ensure that this tracer is also optimized for production code for folks who want to instrument but do not want to send their data off to another system quite yet.

Actually, I revise the above to question explicitly -- for a developer instrumenting but not yet set up with a trace sink location, what options do we plan to expose for them to be able to, for instance, just hook up all trace data to go to standard out (ideally we expose an event so they can send to whatever `out` they wish).

Reply:
This tracer is meant to be used for (unit) testing and definitely not for production. Performance therefore isn't that much of a concern. 

Using `Concurrent` would be nice but it's not enough (as e.g. `_baggage` is a regular variable that is replaced) so we need this locking logic anyway. Right now it's at least consistent and more similar to what Java has. 

> Actually, I revise the above to question explicitly -- for a developer instrumenting but not yet set up with a trace sink location, what options do we plan to expose for them to be able to, for instance, just hook up all trace data to go to standard out (ideally we expose an event so they can send to whatever out they wish).

I guess this would be a good fit for a opentracing-contrib project. Something like this shouldn't be part of the core API library as it probably needs a different release cycle etc.